src/Microsoft.ML.Tokenizers/Model/Tiktoken.cs

```diff
@@ -52,21 +52,58 @@ public sealed class Tiktoken : Model
         {
         }
 
-        internal Tiktoken(
-            Dictionary<ReadOnlyMemory<byte>, int> encoder,
-            IReadOnlyDictionary<int, byte[]> decoder,
-            Dictionary<string, int> vocab,
+        /// <summary>
+        /// Create a new Tiktoken tokenizer object.
+        /// </summary>
+        /// <param name="encoder">The dictionary mapping token utf-8 bytes to Ids.</param>
+        /// <param name="decoder">The dictionary mapping Ids to token utf-8 bytes.</param>
+        /// <param name="vocab">The dictionary mapping string tokens to Ids.</param>
+        /// <param name="specialTokensEncoder">The dictionary mapping special tokens to Ids.</param>
+        /// <param name="cacheSize">The max size of the cache to use.</param>
+        public Tiktoken(
+            IReadOnlyDictionary<ReadOnlyMemory<byte>, int> encoder,
+            IReadOnlyDictionary<int, byte[]>? decoder,
+            IReadOnlyDictionary<string, int>? vocab,
             IReadOnlyDictionary<string, int>? specialTokensEncoder = null,
             int cacheSize = LruCache<string, int[]>.DefaultCacheSize) : this(cacheSize)
         {
-            Debug.Assert(encoder is not null);
-            Debug.Assert(decoder is not null);
-            Debug.Assert(vocab is not null);
+            if (encoder is null)
+            {
+                throw new ArgumentNullException(nameof(encoder));
+            }
+
+            _encoder = encoder;
+
+            if (decoder is null)
+            {
+                decoder = encoder.ToDictionary(kvp => kvp.Value, kvp => kvp.Key.ToArray());
+            }
+
+            if (encoder.Count != decoder.Count)
+            {
+                throw new ArgumentException("The encoder and decoder dictionaries must have the same number of elements");
```

Comment:
I don't think this is possible. The encoder always encode the bytes to a unique Id. round tripping should always work because the bytes and Ids are aways one-to-one mapping. Otherwise the model will be broken. Can you think in example break that?

Reply:
This ctor is public and accepts a dictionary for each. It's why you're checking that the counts are the same. What happens if someone passes in dictionaries with the same counts but don't roundtrip? What prevents that from being possible?  If we allow through such misaligned dictionaries, will anything "bad" happen elsewhere in the implementation, e.g. could we end up in infinite loops or anything like that?

Reply:
We don't do anything round trip so infinite loop is not possible. Does the concern here is to ensure avoiding issues like infinite looping or DOS? or is the concern about users unintentionally can pass bad data? The check we are doing only to help the user to early detect if there is something wrong with the data. 

The other idea we can do here is to not include the decoder in the parameters and we always generate it ourselves. This will be slower and allocation but maybe ok in the scenario we need to support which try avoiding reloading the data from file/stream to create a new Tiktoken object. What do you think?

Reply:
> Does the concern here is to ensure avoiding issues like infinite looping or DOS?

This was my primary concern.

> or is the concern about users unintentionally can pass bad data?

This was my secondary concern, that the API might lead folks into thinking it's ok if they're not mirrors of each other.

> The other idea we can do here is to not include the decoder in the parameters and we always generate it ourselves. This will be slower and allocation but maybe ok in the scenario we need to support which try avoiding reloading the data from file/stream to create a new Tiktoken object. What do you think?

How would someone avoid those overheads? You're thinking of the case where they create the dictionary once and then reuse it for multiple Tiktoken objects, whereas if we generate it ourselves in the ctor, we'd be creating the inverse dictionary each time?  I don't have a strong opinion on that... presumably the 99% case is you only ever create one of these?

Reply:
The scenario in my mind was someone creating an object using the file/stream (the heavy lifting way), then wants to create another object with different special tokens. In that case, you will just get the encoder, decoder, vocab from the first object and pass these to the constructor accept these. My thought was just to use these without copying or recreation. If we think this scenario is unlikely (which I don't know) we can get rid of this constructor all together.

Reply:
I suggest we start without it, and we later add additional overloads that take it if it's deemed valuable as an optimization folks need. We might also come up with different ways to streamline such a scenario when necessary.

Reply:
I made this constructor internal for now (it is used from the factory method in the Tokenizer class). I cleaned up the code there too.